### ISSUE TITLE
RBML layer in CNNs (package) produces NaNs

### ISSUE THREAD

**Issue by josepllberral** [josepllberral/machine-learning-tools#8]
Example:

```
data(mnist)

img_size <- c(28,28);

train <- mnist$train;
training_x <- array(train$x, c(nrow(train$x), 1, img_size)) / 255;
training_y <- binarization(train$y);

test <- mnist$test;
testing_x <- array(test$x, c(nrow(test$x), 1, img_size)) / 255;
testing_y <- binarization(test$y);

dataset <- training_x[1:1000,,,, drop=FALSE];
targets <- training_y[1:1000,, drop=FALSE];

newdata <- testing_x[1:1000,,,, drop=FALSE];

layers <- list(
    c('type' = "CONV", 'n_channels' = 1, 'n_filters' = 4, 'filter_size' = 5, 'scale' = 0.1, 'border_mode' = 'same'),
    c('type' = "POOL", 'n_channels' = 4, 'scale' = 0.1, 'win_size' = 3, 'stride' = 2),
    c('type' = "RELU", 'n_channels' = 4),
    c('type' = "CONV", 'n_channels' = 4, 'n_filters' = 16, 'filter_size' = 5, 'scale' = 0.1, 'border_mode' = 'same'),
    c('type' = "POOL", 'n_channels' = 16, 'scale' = 0.1, 'win_size' = 3, 'stride' = 2),
    c('type' = "RELU", 'n_channels' = 16),
    c('type' = "FLAT", 'n_channels' = 16),
    c('type' = "LINE", 'n_visible' = 784, 'n_hidden' = 64, 'scale' = 0.1),
    c('type' = "RBML", 'n_visible' = 64, 'n_hidden' = 32, 'scale' = 0.1, 'n_gibbs' = 4),
    c('type' = "RELV"),
    c('type' = "LINE", 'n_visible' = 32, 'n_hidden' = 10, 'scale' = 0.1),
    c('type' = "SOFT", 'n_inputs' = 10)
);

mnist_cnn <- train.cnn(dataset, targets, layers, batch_size = 10, training_epochs = 3, learning_rate = 1e-3, momentum = 0.8, rand_seed = 1234);
str(mnist_cnn)

```
Then RBML layer and previous (due to backprop) get NaN values. Tested changing RBML by LINE, and works fine

**Comment by josepllberral** [comment]
Produces NaNs after two non-trivial layers, and this "resets" after a sigmoid.
```
layers <- list(
    c('type' = "LINE", 'n_visible' = 784, 'n_hidden' = 32, 'scale' = 0.1),
    c('type' = "LINE", 'n_visible' = 32, 'n_hidden' = 16, 'scale' = 0.1),
    c('type' = "SIGM", 'n_inputs' = 16),
    c('type' = "RBML", 'n_visible' = 16, 'n_hidden' = 10, 'scale' = 0.1, 'n_gibbs' = 4),
    c('type' = "SOFT", 'n_inputs' = 10)
);
```
Without SIGM produces NaNs.

Also, when first or after a single LINE, this layer behaves OK

**Comment by josepllberral** [comment]
Fixed. There was a misplaced sign in a "dgemm" operation.
